File: app/src/main/java/Controller/PostsRecyclerViewAdapter.java
package Controller;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.example.insta.R;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.Blob;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.FirebaseFirestoreSettings;

import java.util.List;

import Model.Post;
import Model.Profile;

public class PostsRecyclerViewAdapter extends RecyclerView.Adapter<PostsRecyclerViewAdapter.ViewHolder> {

    private final Context context;
    List<Post> posts;
    FirebaseFirestore fStore;
    FirebaseAuth auth;


    public PostsRecyclerViewAdapter(List<Post> posts, Context context) {
        this.posts = posts;
        this.context = context;

        auth = FirebaseAuth.getInstance();
        fStore = FirebaseFirestore.getInstance();
        fStore.setFirestoreSettings(new FirebaseFirestoreSettings.Builder()
                .setPersistenceEnabled(true)
                .setCacheSizeBytes(50000000).build());
    }

    @NonNull
    @Override
    public PostsRecyclerViewAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.post_view_card, parent, false);
        return new PostsRecyclerViewAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder holder, final int position) {
        final Post post = posts.get(position);
        fStore.collection("Users")
                .document(post.getUID())
                .addSnapshotListener(new EventListener<DocumentSnapshot>() {
                    @Override
                    public void onEvent(@Nullable DocumentSnapshot ds, @Nullable FirebaseFirestoreException e) {
                        Profile profile = ds.toObject(Profile.class);
                        holder.profileImage.setImageBitmap(BitmapFactory.decodeByteArray(profile.getProfilePic().toBytes(),
                                0, profile.getProfilePic().toBytes().length));
                        holder.profileName.setText(profile.getName());
                        holder.postImage.setImageBitmap(getPostImageAsBitmap(post.getPostImage()));
                        holder.date.setText(post.getTime().toDate().toString());
                        holder.caption.setText(post.getCaption());
                        holder.love_number.setText(String.valueOf(post.getLikes().size()));
                        holder.love.setImageResource(post.getLiked() ? R.drawable.loved : R.drawable.love);
                        holder.bar.setVisibility(View.GONE);
                    }
                });


        holder.love.setOnClickListener(new View.OnClickListener() {
            @SuppressLint("SetTextI18n")
            @Override
            public void onClick(View v) {
                holder.love.setEnabled(false);
                if (!posts.get(position).getLiked()) {
                    posts.get(position).setLiked(true);
                    posts.get(position).addLike(auth.getUid());
                    holder.love_number.setText(String.valueOf(posts.get(position).getLikes().size()));
                    holder.love.setImageResource(R.drawable.loved);
                    setLike(posts.get(position));
                    holder.love.setEnabled(true);
                } else {
                    posts.get(position).setLiked(false);
                    posts.get(position).removeLike(auth.getUid());
                    holder.love_number.setText(String.valueOf(posts.get(position).getLikes().size()));
                    holder.love.setImageResource(R.drawable.love);
                    removeLike(posts.get(position));
                    holder.love.setEnabled(true);
                }
            }
        });
    }

    public Bitmap getPostImageAsBitmap(Blob postImage) {
        byte[] b = postImage.toBytes();
        Bitmap bitmap = BitmapFactory.decodeByteArray(b, 0, b.length);
        return bitmap;
    }

    private void removeLike(final Post post) {
        fStore.collection("Users")
                .document(post.getUID())
                .collection("Posts")
                .document(post.getID())
                .update("likes", post.getLikes());
    }

    private void setLike(final Post post) {
        fStore.collection("Users")
                .document(post.getUID())
                .collection("Posts")
                .document(post.getID())
                .update("likes", post.getLikes());
    }

    @Override
    public int getItemCount() {
        return posts.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView date, profileName, caption, love_number;
        ImageView love, postImage, profileImage;
        public ProgressBar bar;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            caption = itemView.findViewById(R.id.caption);
            postImage = itemView.findViewById(R.id.post_image);
            love_number = itemView.findViewById(R.id.love_number);
            bar = itemView.findViewById(R.id.progress_bar);
            profileImage = itemView.findViewById(R.id.profile_image);
            date = itemView.findViewById(R.id.date);
            profileName = itemView.findViewById(R.id.profile_name);
            love = itemView.findViewById(R.id.love);
        }
    }
}

File: app/src/main/java/Controller/PostsRecyclerViewAdapter.kt
package Controller

import Model.Date_Time.timeFromNow
import Model.Post
import Model.Profile
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.*
import com.whiteside.insta.R

class PostsRecyclerViewAdapter(var posts: List<Post?>, private val context: Context?) : RecyclerView.Adapter<PostsRecyclerViewAdapter.ViewHolder>() {
    var fStore: FirebaseFirestore
    var auth: FirebaseAuth
    fun update(posts: List<Post?>) {
        this.posts = posts
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.post_view_card, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val post = posts[position]
        fStore.collection("Users")
                .document(post!!.uId!!)
                .addSnapshotListener(EventListener { ds, e ->
                    if (e != null) {
                        e.printStackTrace()
                        return@EventListener
                    }
                    val profile = ds!!.toObject(Profile::class.java)
                    holder.profileImage.setImageBitmap(BitmapFactory.decodeByteArray(profile!!.profilePic!!.toBytes(),
                            0, profile.profilePic!!.toBytes().size))
                    holder.profileName.text = profile.name
                    holder.postImage.setImageBitmap(getPostImageAsBitmap(post.postImage))
                    holder.date.text = timeFromNow(post.time!!)
                    holder.caption.text = post.caption
                    holder.love_number.text = post.likes!!.size.toString()
                    holder.love.setImageResource(if (post.liked!!) R.drawable.loved else R.drawable.love)
                    holder.bar.visibility = View.GONE
                })
        holder.love.setOnClickListener {
            holder.love.isEnabled = false
            if (!post.liked!!) {
                post.liked = true
                post.addLike(auth.uid!!)
                holder.love_number.text = post.likes!!.size.toString()
                holder.love.setImageResource(R.drawable.loved)
                setLike(post)
                holder.love.isEnabled = true
            } else {
                post.liked = false
                post.removeLike(auth.uid!!)
                holder.love_number.text = post.likes!!.size.toString()
                holder.love.setImageResource(R.drawable.love)
                removeLike(post)
                holder.love.isEnabled = true
            }
        }
    }

    fun getPostImageAsBitmap(postImage: Blob?): Bitmap {
        val b = postImage!!.toBytes()
        return BitmapFactory.decodeByteArray(b, 0, b.size)
    }

    private fun removeLike(post: Post?) {
        fStore.collection("Users")
                .document(post!!.uId!!)
                .collection("Posts")
                .document(post.Id!!)
                .update("likes", post.likes)
    }

    private fun setLike(post: Post?) {
        fStore.collection("Users")
                .document(post!!.uId!!)
                .collection("Posts")
                .document(post.Id!!)
                .update("likes", post.likes)
    }

    override fun getItemCount(): Int {
        return posts.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var bar: ProgressBar
        var date: TextView
        var profileName: TextView
        var caption: TextView
        var love_number: TextView
        var love: ImageView
        var postImage: ImageView
        var profileImage: ImageView

        init {
            caption = itemView.findViewById(R.id.caption)
            postImage = itemView.findViewById(R.id.post_image)
            love_number = itemView.findViewById(R.id.love_number)
            bar = itemView.findViewById(R.id.progress_bar)
            profileImage = itemView.findViewById(R.id.profile_image)
            date = itemView.findViewById(R.id.date)
            profileName = itemView.findViewById(R.id.profile_name)
            love = itemView.findViewById(R.id.love)
        }
    }

    init {
        auth = FirebaseAuth.getInstance()
        fStore = FirebaseFirestore.getInstance()
    }
}